TUNEBI-33 Configurable Shard

haszhixuan opened the pull request:
Due to the reason that for configurable shard setting, There will be frequent S3 requests on very small files. The requests could fail with many reasons: connection closed by peer, read timeout and etc. Sometimes it failed the default 10 times and the process exits out. 

Fixes include:
1. Changed the retry to 20 times;
2. Added an incremental timeout in between each retry; 
3. Added more debug print for easier investigation.

The other format changes are due to go fmt.

haszhixuan committed:
TUNEBI-33 Configurable Shard

Due to the reason that for configurable shard setting, There will be frenquent S3 requests on very small files. The request could fail with many reasons:
connection closed by peer, read timeout and etc. Sometimes it failed the default 10 times and the process exits out. Changed the timeout try to 20 times;
Added an incremental timeout in between each try; and added more debug print for easier investigation.

hasnickl commented on getter.go:
```diff
@@ -108,19 +107,26 @@ func newGetter(getURL url.URL, c *Config, b *Bucket) (io.ReadCloser, http.Header
-		if err == nil {
+
+		// OK status
+		if err == nil && resp != nil && resp.StatusCode == 200 {
```
i dont think we should add a StatusCode check here since callers are already checking status code

EDIT: at minimum, we should only check that 200 <= StatusCode < 300

mhoglan replied:
i know on the other we check for 206 since we do chunked requests.  here we doing single requests.  there are reasons you would get 2XX response codes but we may just not see them currently.   We can keep this checking for just 200 but I would add a comment to the top of the function specifying that explicit requirement that we will retry till we get a 200 OK

hasnickl replied:
i checked the chunked requests and those use an HTTP client directly, they do not go through this method

haszhixuan replied:
I think the check for 200 here is to make sure this is a completely successful request (no err, non nil respond and OK status code) and we should return without retrying. I can add more comments there.

haszhixuan commented on getter.go:
```diff
@@ -132,11 +138,15 @@ func (g *getter) queueFile(url *url.URL) (http.Header, error) {
+		logger.debugPrintf("ERROR on queueFile", errgo.Mask(err))
+		return nil, err
 	}
 
```
This change actually avoids accessing resp (removing line 136) when err is not nil (I believe we had crashed here before). I can add more checking to return resp.header when resp is not nil, return nil otherwise.

haszhixuan commented on getter.go:
```diff
@@ -108,19 +107,26 @@ func newGetter(getURL url.URL, c *Config, b *Bucket) (io.ReadCloser, http.Header
 		if err != nil {
-			return
+			logger.debugPrintf("NewRequest error on attempt %d: retrying url: %s, error: %s", i, urlStr, err)
+			continue // let retry
```
This makes sense. I looked into NewRequest() and see it does not talk to S3 at all. I will change it back to return.

haszhixuan committed:
TUNEBI-33

Added more conments for some conditional blocks.
Skip retrying for NewRequest failure because it does not involve communcation with S3.
